File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size()==0){
            return {};
        }
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        for(int i = 0; i<nums.size(); i++){
            if(nums[i]>0){
                break;
            }
            while(i>0 && i<nums.size() && nums[i]==nums[i-1]){
                i++;
            }
            int l = i+1;
            int r = nums.size()-1;
            while(l<r){
                if(nums[i] + nums[l] + nums[r]>0){
                    r--;
                } else if(nums[i] + nums[l] + nums[r]<0){
                    l++;
                } else{
                    vector<int> res = {nums[i], nums[l], nums[r]};
                    ans.push_back(res);
                    l++;
                    r--;
                    while(l<r && nums[l]==nums[l-1]){
                        l++;
                    }
                }
            }
        }
        return ans;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        res = []
        nums.sort()

        for i, a in enumerate(nums):
            if a>0:
                break
            
            if i>0 and nums[i-1] == nums[i]:
                continue
            
            l = i+1
            r = len(nums)-1
            while l<r:
                sum = a+nums[l]+nums[r]
                if sum>0:
                    r-=1
                elif sum<0:
                    l+=1
                else:
                    res.append([a, nums[l], nums[r]])
                    l+=1
                    r-=1
                    while l<r and nums[l-1] == nums[l]:
                        l+=1
        return res